src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlTypes/SqlTypeWorkarounds.netfx.cs

```diff
@@ -0,0 +1,342 @@
+                // assumptions. Importantly, the type must be [Serializable] but cannot be ISerializable, as the
+                // presence of the interface means that the type wants to be responsible for its own serialization,
+                // and that member fields are not guaranteed to be part of the serialization contract. Additionally,
+                // we need to check for [OnSerializing] and [OnDeserializing] methods, because we cannot account
+                // for any logic which might be present within them.
+
+                if (!typeof(SqlDecimal).IsSerializable)
+                {
+                    SqlClientEventSource.Log.TryTraceEvent("SqlTypeWorkarounds.SqlDecimalHelper.GetFastDecomposer | Info | SqlDecimal isn't Serializable. Less efficient fallback method will be used.");
+                    return null; // type is not serializable - cannot use fast path assumptions
+                }
+
+                if (typeof(ISerializable).IsAssignableFrom(typeof(SqlDecimal)))
+                {
+                    SqlClientEventSource.Log.TryTraceEvent("SqlTypeWorkarounds.SqlDecimalHelper.GetFastDecomposer | Info | SqlDecimal is ISerializable. Less efficient fallback method will be used.");
+                    return null; // type contains custom logic - cannot use fast path assumptions
+                }
+
+                foreach (MethodInfo method in typeof(SqlDecimal).GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
+                {
+                    if (method.IsDefined(typeof(OnDeserializingAttribute)) || method.IsDefined(typeof(OnDeserializedAttribute)))
+                    {
+                        SqlClientEventSource.Log.TryTraceEvent("SqlTypeWorkarounds.SqlDecimalHelper.GetFastDecomposer | Info | SqlDecimal contains custom serialization logic. Less efficient fallback method will be used.");
+                        return null; // type contains custom logic - cannot use fast path assumptions
+                    }
+                }
+
+                // GetSerializableMembers filters out [NonSerialized] fields for us automatically.
+
+                FieldInfo fiData1 = null, fiData2 = null, fiData3 = null, fiData4 = null;
+                foreach (MemberInfo candidate in FormatterServices.GetSerializableMembers(typeof(SqlDecimal)))
+                {
+                    if (candidate is FieldInfo fi && fi.FieldType == typeof(uint))
+                    {
+                        if (fi.Name == "m_data1")
+                        { fiData1 = fi; }
+                        else if (fi.Name == "m_data2")
+                        { fiData2 = fi; }
+                        else if (fi.Name == "m_data3")
+                        { fiData3 = fi; }
+                        else if (fi.Name == "m_data4")
+                        { fiData4 = fi; }
+                    }
```

Comment:
Do we need if statements here instead of else if? since we return null if any of them is null on line 195.

Reply:
No. fi.Name can be only one value at a time inside the loop so if we get a match on any one, we don't need to process any of the other conditions.